### DISCUSSION TITLE
Reference handling in behavior callbacks

### DISCUSSION THREAD

**Discussion by dominiklohmann** [actor-framework/actor-framework#1220] [Q&A] [answered]
I was wondering recently how value categories and constness of parameters affect behavior callbacks combined with message uniqueness. My current understanding is the following:

```c++
{
  [](T x) {
    // Always copies from the message into x.
  },
  [](T& x) {
    // Passes a reference from the message to x if the message is not shared,
    // otherwise copies. This allows for moving out of the message.
  },
  [](T const& x) {
    // Always passes a reference from the message to x.
  },
  [](T&& x) {
    // Does not compile.
  },
  [](T const&& x) {
    // Nonsensical; does not compile.
  },
}
```

This leaves me with three open questions:
1. Is my understanding correct?
2. Should this be documented prominently in the [Message Passing documentation](https://actor-framework.readthedocs.io/en/stable/MessagePassing.html)? To me, this feels like something that should be fundamental knowledge.
3. Would it be possible to support the non-const rvalue-reference case by failing to match the message if it is not unique?

**Comment by Neverlord** [comment, marked as answer]
>     1. Is my understanding correct?

Yep. Maybe one small addition for the sake of completeness. Consider this handler: `[](A&, B&, const C&)`. CAF looks at the signature of a message handler and checks whether *any* argument takes a mutable reference. If so, CAF forces the message to “unshare". This is a no-op if the reference count equals 1. Otherwise, unsharing forces a deep copy (and then the reference count of the copy is always 1).

So there’s no copies for individual elements. If you inspect the current message (e.g. via `current_mailbox_element()->payload`), you will always find a message with reference count 1 when inside a message handler that takes mutable references.

>     2. Should this be documented prominently in the [Message Passing documentation](https://actor-framework.readthedocs.io/en/stable/MessagePassing.html)? To me, this feels like something that should be fundamental knowledge.

I think you’re right. That section looks like it could be improved.

>     3. Would it be possible to support the non-const rvalue-reference case by failing to match the message if it is not unique?

Technically possible: yes. But I'd be hesitant to add such a feature. I think that's a feature that would add a lot of complexity and make it all too easy to shoot yourself in the foot. I'd also imagine there are few use cases.

You can't "not match" inside a message handler, but if you only wish this as an optimization technique: supporting that would be straightforward. Earlier versions of CAF had a `current_message()` function on actors. We could bring that back to allow something like this:

```cpp
[self](const my_data& x) {
  if (self->current_message().unique()) {
    auto& mutable_x = self->current_message().get_mutable_as<my_data>(0);
    // ...
  } else {
    // ...
  }
}
```

**Reply by dominiklohmann** [reply]
Thanks! That is actually really helpful. I don't need the `current_message` functionality, I mostly wanted clarification whether my understanding was correct.

Regarding the non-const rvalue-reference, I agree that should not be a thing.

I wonder if there's room for optimization for partially unshared messages with the flat message design of CAF 0.18, and making the check against each argument individually.
